### PR TITLE
Remove xenial references

### PR DESCRIPTION
**Pull request by aciba90**
## Proposed Commit Message

```
Remove Xenial references

- Remove references and dead code to Xenial, Eoan, Python < 3.7
- cc_ubuntu_drivers: Use python3-debconf instead of shell script
```

## Additional Context
SC-710

## Test Steps

To test the cc_ubuntu_drivers changes:
```sh
cat << EOT > /tmp/my-user-data
#cloud-config
drivers:
  nvidia:
    license-accepted: true
EOT

lxc launch --vm ubuntu:focal un --config=user.user-data="$(cat /tmp/my-user-data)"
```

## Checklist:

 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/contributing.html)
 - [x] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any documentation accordingly

### CODE REVIEW

**Comment by aciba90** [comment]
I have some concerns related to this changes:

- With the changes in `cc_ubuntu_drivers` we now depend on `python3-debconf`, but I was not sure how to add it as a dependency for debian, rpm, etc packages
- Should we add an integration test for `cc_ubuntu_drivers`?
-  I was not sure about what to do in [cc_puppet.py#L260](https://github.com/canonical/cloud-init/blob/24293f2a1a72a2233c879ae74e64b5008123ac14/cloudinit/config/cc_puppet.py#L260), maybe @blackboxsw cloud clarify it.

Thank you.

File: cloudinit/config/cc_ubuntu_drivers.py
Comment (blackboxsw) [inline cloudinit/config/cc_ubuntu_drivers.py:105]: We probably do not need this check here, we only need to ensure we add PRs which place `python3-debconf` in the ubuntu/bionic|focal|impish|devel branches `debian/control` files and `packages/debian/control.in` in this branch.

From an RPM perspective and debian-proper, they won't care to pull in this dependency as ubuntu_drivers module only runs on Ubuntu.

File: doc/rtd/topics/network-config-format-v2.rst
Comment (blackboxsw) [inline doc/rtd/topics/network-config-format-v2.rst:341]: confirmed fix-released in [netplan bug](https://bugs.launchpad.net/netplan/+bug/1756701) and in (upstream netplan source](https://github.com/canonical/netplan/blob/main/src/parse.c#L2001)

File: packages/bddeb
Comment (blackboxsw) [inline packages/bddeb:42]: Let's make this
```suggestion
    >>> get_release_suffix("jammy")
    '~22.04.1'
```

**Comment by blackboxsw** [comment]
> I have some concerns related to this changes:
> 
> * With the changes in `cc_ubuntu_drivers` we now depend on `python3-debconf`, but I was not sure how to add it as a dependency for debian, rpm, etc packages
> * Should we add an integration test for `cc_ubuntu_drivers`?

Now that pycloudlib officially supports Oracle cloud launches we might be able to cobble up an Oracle specific integration test which would require providing the ubuntu: drivers directives

See more info with examples of driving this on oracle:
https://ubuntu.com/blog/installing-nvidia-drivers-on-oracle-cloud


> * I was not sure about what to do in [cc_puppet.py#L260](https://github.com/canonical/cloud-init/blob/24293f2a1a72a2233c879ae74e64b5008123ac14/cloudinit/config/cc_puppet.py#L260), maybe @blackboxsw cloud clarify it.

You can drop that comment safely now. I added it in commit 4fe40704bf but we ended up fixing the actual line below a few months later in commit fecbd81889 and forgot to drop the comment.

File: cloudinit/config/cc_ubuntu_drivers.py
Comment (aciba90) [reply cloudinit/config/cc_ubuntu_drivers.py:105]: Removed. Thanks for the clarification. I will make those PRs after this one is merged.

File: config/cloud.cfg.tmpl
Comment (aciba90) [reply config/cloud.cfg.tmpl:212]: Done

File: packages/bddeb
Comment (aciba90) [reply packages/bddeb:42]: Done

**Comment by aciba90** [comment]
> You can drop that comment safely now. I added it in commit [4fe4070](https://github.com/canonical/cloud-init/commit/4fe40704bfb998588d6a96b4a2773ddc9bd692ac) but we ended up fixing the actual line below a few months later in commit [fecbd81](https://github.com/canonical/cloud-init/commit/fecbd81889011e8a75badd18935297f3494fe485) and forgot to drop the comment.

Dropped

> Now that pycloudlib officially supports Oracle cloud launches we might be able to cobble up an Oracle specific integration test which would require providing the ubuntu: drivers directives

Thanks for pointing out which instance type would be better for doing this test. I am going to work on that integration test.

**Comment by aciba90** [comment]
Reopen to retrigger Travis CI.

**Comment by aciba90** [comment]
Integration test added. Ready to be reviewed.

File: config/cloud.cfg.tmpl
Comment (TheRealFalcon) [reply config/cloud.cfg.tmpl]: https://github.com/canonical/cloud-init/blob/ubuntu/bionic/debian/patches/renderer-do-not-prefer-netplan.patch#L14

File: cloudinit/debconf.py
Comment (blackboxsw) [inline cloudinit/debconf.py:1]: @aciba90 I don't think we really need to surface this top-level module yet, because our only consumer of this functionality in this PR is `cloudinit.config.cc_ubuntu_drivers` if we grow support for debconf use in other modules (like cc_apt*) then I could see us maybe needing this level of ImportError guarding in a top-level module definition. As it is, the only callers of this logic will be cc_ubuntu_drivers, so let's to the try/except ImportError dance in that module specifically. Since Ubuntu packaging will define a strict `Depends: python3-debconf` we won't be concerned about the absence of that import. Any distros will not otherwise need python3-debconf and so the ImportError handling can just pass in those cases.

File: config/cloud.cfg.tmpl
Comment (aciba90) [reply config/cloud.cfg.tmpl]: Removed, thanks. I did not know that.

File: cloudinit/debconf.py
Comment (aciba90) [reply cloudinit/debconf.py]: `debconf` import moved to `cloudinit.config.cc_ubuntu_drivers` and unit tested. Thanks.

File: config/cloud.cfg.tmpl
Comment (blackboxsw) [reply config/cloud.cfg.tmpl]: Meh, I was holding off on landing this branch until dust settled for SRU specifically thinking in terms of that daily recipe drift that would result. But sure, not touching it makes our lives easier. +1.

File: tests/integration_tests/modules/test_ubuntu_drivers.py
Comment (blackboxsw) [inline tests/integration_tests/modules/test_ubuntu_drivers.py:18]: Good integration test, looks reasonable. I can't seem to get it to succeed locally though. I don't think my region has VM.GPU3.1 availability though `availability_domain = "qIZq:PHX-AD-1" ` I'm also seeing errors from Oracle cloud dashboard UI
```
This shape is either not compatible with the selected image, or not available in the current availability domain.
```
I'll try sorting this with you tomorrow.

File: cloudinit/config/cc_ubuntu_drivers.py
Comment (aciba90) [reply cloudinit/config/cc_ubuntu_drivers.py]: Ups, fixed, thanks!

File: cloudinit/config/cc_ubuntu_drivers.py
Comment (aciba90) [reply cloudinit/config/cc_ubuntu_drivers.py]: Make sense. Simplified, thanks.

File: tests/integration_tests/modules/test_ubuntu_drivers.py
Comment (aciba90) [reply tests/integration_tests/modules/test_ubuntu_drivers.py:18]: True. I tried to run it with my credentials, but they do not have permission to run this instance shape. I assumed:

* We need a shape containing an Nvidia card: [vmshapes__vm-gpu](https://docs.oracle.com/en-us/iaas/Content/Compute/References/computeshapes.htm#vmshapes__vm-gpu).
* The CI OCI credentials is allowed to instantiate this kind of shape.

If there is some way I can help, please let me know.

Comment (blackboxsw) [reply tests/integration_tests/modules/test_ubuntu_drivers.py:19]: Turns out is was a couple of pycloudlib bugs for oracle that broke me
- no jammy support: https://github.com/canonical/pycloudlib/pull/200
- inability to choose the right subnet based on availability_domain: https://github.com/canonical/pycloudlib/pull/201


With those issues fixed I could work through your integration test to finalize the review.

File: tests/integration_tests/modules/test_ubuntu_drivers.py
Comment (blackboxsw) [inline tests/integration_tests/modules/test_ubuntu_drivers.py:28]: This command will fail because invoking apt from a script exits non-zero saying something like `don't use apt interfacefrom scripts`.

Here's an alternative (that I put into the full diff on my final comment 
```suggestion
        result = client.execute("dpkg -l | grep nvidia")
        assert result.ok, "No nvidia packages found"
        assert re.search(
            r"ii\s+linux-modules-nvidia-\d+-server", result.stdout
        ), (
            f"Did not find specific nvidia drivers packages in:"
            f" {result.stdout}"
         )
```

**Review by blackboxsw — APPROVED** [review]
Looks good. with pycloudlib related fixes landed and the following patch below applied I can confirm this integration test is in good shape.
```diff
diff --git a/tests/integration_tests/modules/test_ubuntu_drivers.py b/tests/integration_tests/modules/test_ubuntu_drivers.py
index 6cf276309..4fbfba3c1 100644
--- a/tests/integration_tests/modules/test_ubuntu_drivers.py
+++ b/tests/integration_tests/modules/test_ubuntu_drivers.py
@@ -1,3 +1,5 @@
+import re
+
 import pytest
 
 from tests.integration_tests.clouds import IntegrationCloud
@@ -10,19 +12,26 @@ drivers:
         license-accepted: true
 """
 
+# NOTE(VM.GPU2.1 is not in all availability_domains: use qIZq:US-ASHBURN-AD-1)
+
 
-@pytest.mark.user_data(USER_DATA)
 @pytest.mark.adhoc  # Expensive instance type
 @pytest.mark.oci
 def test_ubuntu_drivers_installed(session_cloud: IntegrationCloud):
     with session_cloud.launch(
-        launch_kwargs={"instance_type": "VM.GPU3.1"}
+        launch_kwargs={"instance_type": "VM.GPU2.1"}, user_data=USER_DATA
     ) as client:
         log = client.read_from_file("/var/log/cloud-init.log")
         verify_clean_log(log)
         assert 1 == log.count(
             "Installing and activating NVIDIA drivers "
-            '("nvidia/license-accepted"=True, version=latest)'
+            "(nvidia/license-accepted=True, version=latest)"
+        )
+        result = client.execute("dpkg -l | grep nvidia")
+        assert result.ok, "No nvidia packages found"
+        assert re.search(
+            r"ii\s+linux-modules-nvidia-\d+-server", result.stdout
+        ), (
+            f"Did not find specific nvidia drivers packages in:"
+            f" {result.stdout}"
         )
-        cmd = "apt list --installed | grep -E 'nvidia-*-server'"
-        assert client.execute(cmd).ok
```

**Comment by aciba90** [comment]
Integration test updated. Thanks a lot, @blackboxsw, for investigating why the shape did not run, fixing `pycloudlib` bugs and the integration test.

Ready to merge after CI.

File: cloudinit/sources/DataSourceOpenStack.py
Comment (aciba90) [reply cloudinit/sources/DataSourceOpenStack.py]: Done, I also removed another possible conflict in `cloudinit/sources/DataSourceEc2.py`.

**Comment by blackboxsw** [comment]
@aciba90 I resolved the merge conflict that as introduced when another PR bumped to an even later comitish of pycloudlib. Running integraiton tests against OCI works great once grabbing git+https://github.com/canonical/pycloudlib.git@6eee33c9c4f630bc9c13b6e48f9ab36e7fb79ca6

**Pull request merged**
